RIOT-OS/RIOT
# net: add Asymcute, an asynchronous MQTT-SN client implementation

### Contribution description
This PR adds a new `MQTT-SN` client implementation called `Asymcute`. In contrary to the exising `emCute` implementation, this one is completely asynchronous, thus allowing for a higher throughput and for more flexibility to the cost of increased memory usage (both RAM and ROM).

`Asymcute` is not only able to handle multiple pending requests concurrently, it is also able to connect to more than one gateway at a time.

The motivation for implementing a second MQTT-SN client comes mainly from some application layer protocol experiments that we were running earlier this year in the context of our 'I3' research project, where we were measuring certain metrics for MQTT-SN, CoAP and some ICN-based approaches. For this, the existing synchronous `emCute` library was not well suited, so hence the need for a fully asynchronous MQTT-SN client library.

### Testing
Easiest way (using native):
- run a local mqtt-sn broker using the build in `rsmb` make target (see #9459)
- run the `examples/asymcute_mqttsn` application under `native`
- play, also hard-kill (`kill -9`) the `rsmb` instance to see the effect due to time out on keep-alive messages. For this it makes sense though to reduce the keep alive interval (override ASYMCUTE_KEEPALIVE).

### Issues/PRs references
rebased/depending on #9459
